sdks/go/examples/wasm/wasm.go

```diff
@@ -0,0 +1,159 @@
+	flag.Parse()
+	beam.Init()
+	ctx := context.Background()
+	if err := preRun(); err != nil {
+		panic(err)
+	}
+	if err := run(ctx); err != nil {
+		panic(err)
+	}
+}
+
+func run(ctx context.Context) error {
+	p, s := beam.NewPipelineWithRoot()
+
+	in := beam.Create(s, "a", "b", "c", "d", "e", "f", "g", "h")
+
+	out := beam.ParDo(s.Scope("embeddedWasm"), &embeddedWasmFn{}, in)
+
+	textio.Write(s, *output, out)
+
+	if err := beamx.Run(ctx, p); err != nil {
+		return fmt.Errorf("failed to run pipeline: %v", err)
+	}
+	return nil
+}
+
+// Concept #2 wrap wasm function execution within a DoFn.
+// wasmFn wraps a DoFn to execute a Rust compiled wasm function
+type embeddedWasmFn struct {
+	r                              wazero.Runtime
+	mod                            api.Module
+	greeting, allocate, deallocate api.Function
+}
+
+// Setup loads and initializes the embedded wasm functions
+// Concept #1: Load a compiled wasm file []byte content and function.
+// This example is derived from
+// https://github.com/tetratelabs/wazero/blob/v1.0.0-pre.3/examples/allocation/rust/greet.go
+func (fn *embeddedWasmFn) Setup(ctx context.Context) error {
+	fn.r = wazero.NewRuntime(ctx)
+	_, err := fn.r.NewHostModuleBuilder("env").
+		NewFunctionBuilder().WithFunc(logString).Export("log").
+		Instantiate(ctx, fn.r)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate host module: %w", err)
+	}
+	fn.mod, err = fn.r.InstantiateModuleFromBinary(ctx, greetWasm)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate wasm module: %v", err)
+	}
+	fn.greeting = fn.mod.ExportedFunction(wasmFunctionName)
+	fn.allocate = fn.mod.ExportedFunction(wasmAllocateFunctionName)
+	fn.deallocate = fn.mod.ExportedFunction(wasmDeallocateFunctionName)
+	return nil
+}
+
+// ProcessElement processes a string calling a wasm function written in Rust
+// This example is derived from
+// https://github.com/tetratelabs/wazero/blob/v1.0.0-pre.3/examples/allocation/rust/greet.go
+func (fn *embeddedWasmFn) ProcessElement(ctx context.Context, s string) (string, error) {
+	size := uint64(len(s))
+	results, err := fn.allocate.Call(ctx, size)
+	if err != nil {
+		return "", fmt.Errorf("error calling allocate: %w", err)
+	}
+	ptr := results[0]
+	defer fn.deallocate.Call(ctx, ptr, size)
+	if !fn.mod.Memory().Write(ctx, uint32(ptr), []byte(s)) {
+		return "", fmt.Errorf("Memory.Write(%d, %d) out of range of memory size %d",
+			ptr, size, fn.mod.Memory().Size(ctx))
+	}
+
+	ptrSize, err := fn.greeting.Call(ctx, ptr, size)
+	resultPtr := uint32(ptrSize[0] >> 32)
+	resultSize := uint32(ptrSize[0])
+	defer fn.deallocate.Call(ctx, uint64(resultPtr), uint64(resultSize))
+	bytes, ok := fn.mod.Memory().Read(ctx, resultPtr, resultSize)
+	if !ok {
+		return "", fmt.Errorf("Memory.Read(%d, %d) out of range of memory size %d",
+			resultPtr, resultSize, fn.mod.Memory().Size(ctx))
+	}
+	return string(bytes), nil
+}
+
+func logString(ctx context.Context, m api.Module, offset, byteCount uint32) {
```

Comment:
Please add a comment to this saying that it's being used as a foreign function to the wasm code. It's easy to miss that when reading the code, and it's not used elsewhere.